snapcraft/internal/lxd.py

```diff
@@ -72,11 +72,14 @@ def __init__(self, *, output, source, project_options,
             kernel = server_environment['kernel_architecture']
         except KeyError:
             kernel = server_environment['kernelarchitecture']
-        deb_arch = _get_deb_arch(kernel)
-        if not deb_arch:
-            raise ContainerConnectionError(
-                'Unrecognized server architecture {}'.format(kernel))
-        self._image = 'ubuntu:xenial/{}'.format(deb_arch)
+        if image is None:
+            deb_arch = _get_deb_arch(kernel)
+            if not deb_arch:
+                raise ContainerConnectionError(
+                    'Unrecognized server architecture {}'.format(kernel))
+            self._image = 'ubuntu:xenial/{}'.format(deb_arch)
+        else:
+            self._image = image
```

Comment:
This provides the ability to build not only on a different release, but a completely different arch. Is that intended?

Reply:
This is true and indeed will probably break horribly. It will also let you specify a fedora container which probably won't work either. But otoh it provides an easy mapping from what you pass to lxd launch to what you pass here. If the user does something silly and it breaks, they get to keep both pieces.

My original version just allowed "ubuntu:$series/$deb_arch" and had a --series argument, but that didn't actually work for my use case of building on artful because artful is only available on the ubuntu-daily: remote, not ubuntu:.

Reply:
Fair enough. I have no strong opinions about the arch, and I think the overall functionality is nice to have. I'll let the others weigh in.

Reply:
Specifying the arch is optional afair - how about checking that the arch is supported by the server if it was specified and a nice error otherwise?

Reply:
I'm sorry, I don't understand what you mean here.

Reply:
In this method `deb_arch` is the image architecture ie. `/amd64` mapped from the kernel architecture reported by the LXD server. We could 1. add that suffix if the image name is eg. `ubuntu:zesty` 2. or check a given image name eg. `ubuntu:zesty/amd64` to see if the last component is an architecture supported by the server.

Reply:
lxd appears to already to do 1. (e.g. lxc launch ubuntu:artful works).

I admit I don't really get the desire to protect the user from themselves here. If I pass something stupid, I assume it fails in a pretty obvious way. Maybe I should try that...

Reply:
Here's what happens:

{{{
(mwhudson/scrollbars)mwhudson@aeglos:~/src/subiquity$ snapcraft cleanbuild --image ubuntu:artful/ppc64el
"grade" property not specified: defaulting to "stable"
Creating snapcraft-songfully-corneous-pearl
error: Requested architecture isn't supported by this host
Traceback (most recent call last):
  File "/snap/snapcraft/x5/bin/snapcraft", line 11, in <module>
    load_entry_point('snapcraft==2.34', 'console_scripts', 'snapcraft')()
  File "/snap/snapcraft/x5/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 565, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/snap/snapcraft/x5/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/snap/snapcraft/x5/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/snap/snapcraft/x5/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/cli/__main__.py", line 19, in <module>
    run(prog_name='snapcraft')
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/cli/lifecycle.py", line 196, in cleanbuild
    lifecycle.cleanbuild(project_options, remote, image)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/internal/lifecycle.py", line 337, in cleanbuild
    metadata=config.get_metadata(), remote=remote, image=image).execute()
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/internal/lxd.py", line 149, in execute
    with self._ensure_started():
  File "/snap/snapcraft/x5/usr/lib/python3.6/contextlib.py", line 82, in __enter__
    return next(self.gen)
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/internal/lxd.py", line 133, in _ensure_started
    self._ensure_container()
  File "/snap/snapcraft/x5/lib/python3.6/site-packages/snapcraft/internal/lxd.py", line 121, in _ensure_container
    'lxc', 'launch', '-e', self._image, self._container_name])
  File "/snap/snapcraft/x5/usr/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['lxc', 'launch', '-e', 'ubuntu:artful/ppc64el', 'local:snapcraft-songfully-corneous-pearl']' returned non-zero exit status 1.
}}}

This seems clear enough to me? It would be better if snapcraft didn't dump the traceback out but that's another issue...

Reply:
That's exactly what I'd like to avoid here. The traceback is what happens by default on unhandled errors.

Reply:
So the fix for that is to handle lxc launch failing better? There are other reasons it might fail (running out of disk space being an obvious one, I guess).

Reply:
That's how we handle it because the return code from the `lxc` command doesn't reflect the type of failure.

Reply:
Thinking on it a little more, I guess you were suggesting to disable the traceback if the command failed at all? That's actually a good idea!
So: we can still avoid the actual lxc error here. But have another PR to catch `lxc init`, `lxc launch` and `lxc exec` respectively failing and (use a SnapcraftError subclass to) provide a clean error message.